src/pyodide/internal/python.ts

```diff
@@ -50,22 +50,50 @@ function prepareWasmLinearMemory(Module: Module): void {
     // The effects of these are purely in Python state so they only need to be run
     // if we didn't restore a snapshot.
     entropyBeforeTopLevel(Module);
-    adjustSysPath(Module);
+    // Note that setupPythonSearchPath runs after adjustSysPathOnlyFor0260 and rearranges where
+    // the /session/metadata path is added.
+    adjustSysPathOnlyFor0260(Module);
   }
   if (Module.API.version !== '0.26.0a2') {
     finalizeBootstrap(Module);
   }
 }
 
-function maybeAddVendorDirectoryToPath(pyodide: Pyodide): void {
+function setupPythonSearchPath(pyodide: Pyodide): void {
+  const site_packages = pyodide._module.FS.sessionSitePackages;
   pyodide.runPython(`
     def _tmp():
       import sys
       from pathlib import Path
 
       VENDOR_PATH = "/session/metadata/vendor"
-      if Path(VENDOR_PATH).is_dir():
-        sys.path.append(VENDOR_PATH)
+      SESSION_PATH = "/session/metadata"
+      SITE_PACKAGES_PATH = "${site_packages}"
+
+      # adjustSysPathOnlyFor0260 adds the session path, but it is immortalised by the memory snapshot. This
+      # code runs irrespective of the memory snapshot.
+      if SESSION_PATH in sys.path:
+        sys.path.remove(SESSION_PATH)
+      if VENDOR_PATH in sys.path:
+        sys.path.remove(VENDOR_PATH)
+      if SITE_PACKAGES_PATH not in sys.path:
+        sys.path.append(SITE_PACKAGES_PATH)
+
+      # Insert vendor path after system paths but before site-packages
+      # System paths are typically: ['/session', '/lib/python312.zip', '/lib/python3.12', '/lib/python3.12/lib-dynload']
+      # We want to insert before '/lib/python3.12/site-packages' and other site-packages
+      #
+      # We also need the session path to be before the vendor path, if we don't do so then a local
+      # import will pick a module from the vendor path rather than the local path. We've got a test
+      # that reproduces this (vendor_dir).
+      for i, path in enumerate(sys.path):
+        if 'site-packages' in path:
+          sys.path.insert(i, VENDOR_PATH)
+          sys.path.insert(i, SESSION_PATH)
```

Comment:
Nit: I'd suggest:
```suggestion
          sys.path[i:i] = [SESSION_PATH, VENDOR_PATH]
```
which is a bit clearer.